bfgs.go

```diff
@@ -59,65 +55,62 @@ func (b *BFGS) Iterate(loc *Location) (Operation, error) {
 func (b *BFGS) InitDirection(loc *Location, dir []float64) (stepSize float64) {
 	dim := len(loc.X)
 	b.dim = dim
+	b.first = true
 
-	b.x = resize(b.x, dim)
-	copy(b.x, loc.X)
-	b.grad = resize(b.grad, dim)
-	copy(b.grad, loc.Gradient)
+	x := mat64.NewVector(dim, loc.X)
+	grad := mat64.NewVector(dim, loc.Gradient)
+	b.x.CloneVec(x)
+	b.grad.CloneVec(grad)
 
-	b.y = resize(b.y, dim)
-	b.s = resize(b.s, dim)
-	b.tmp = resize(b.tmp, dim)
-	b.yVec = mat64.NewVector(dim, b.y)
-	b.sVec = mat64.NewVector(dim, b.s)
-	b.tmpVec = mat64.NewVector(dim, b.tmp)
+	b.y.Reset()
+	b.s.Reset()
+	b.tmp.Reset()
 
 	if b.invHess == nil || cap(b.invHess.RawSymmetric().Data) < dim*dim {
 		b.invHess = mat64.NewSymDense(dim, nil)
 	} else {
 		b.invHess = mat64.NewSymDense(dim, b.invHess.RawSymmetric().Data[:dim*dim])
 	}
-
-	// The values of the hessian are initialized in the first call to NextDirection
-
-	// initial direcion is just negative of gradient because the hessian is 1
-	copy(dir, loc.Gradient)
-	floats.Scale(-1, dir)
-
-	b.first = true
-
-	return 1 / floats.Norm(dir, 2)
+	// The values of the inverse Hessian are initialized in the first call to
+	// NextDirection.
+
+	// Initial direction is just negative of the gradient because the Hessian
+	// is an identity matrix.
+	d := mat64.NewVector(dim, dir)
+	d.ScaleVec(-1, grad)
+	return 1 / mat64.Norm(d, 2)
 }
 
 func (b *BFGS) NextDirection(loc *Location, dir []float64) (stepSize float64) {
-	if len(loc.X) != b.dim {
+	dim := b.dim
+	if len(loc.X) != dim {
 		panic("bfgs: unexpected size mismatch")
 	}
-	if len(loc.Gradient) != b.dim {
+	if len(loc.Gradient) != dim {
 		panic("bfgs: unexpected size mismatch")
 	}
-	if len(dir) != b.dim {
+	if len(dir) != dim {
 		panic("bfgs: unexpected size mismatch")
 	}
 
-	// Compute the gradient difference in the last step
-	// y = g_{k+1} - g_{k}
-	floats.SubTo(b.y, loc.Gradient, b.grad)
+	x := mat64.NewVector(dim, loc.X)
+	grad := mat64.NewVector(dim, loc.Gradient)
 
-	// Compute the step difference
 	// s = x_{k+1} - x_{k}
-	floats.SubTo(b.s, loc.X, b.x)
+	b.s.SubVec(x, &b.x)
+	// y = g_{k+1} - g_{k}
+	b.y.SubVec(grad, &b.grad)
 
-	sDotY := floats.Dot(b.s, b.y)
-	sDotYSquared := sDotY * sDotY
+	sDotY := mat64.Dot(&b.s, &b.y)
```

Comment:
Could you please open up an issue and/or PR to fast path Vector in Dot? We'd like to keep adding to the efficiency of mat64, and us needing it is a particularly good thing to add.